docs/source/extensions/loops.rst

```diff
@@ -0,0 +1,382 @@
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+What remains in the Trainer is the loop, :code:`zero_grad()`, :code:`backward()` and :code:`optimizer.step()` calls.
+These are considered *boilerplate* and get automated by Lightning.
+
+This optimization scheme is very general and applies to the vast majority of deep learning research today.
+However, the loops and optimizer calls here remain predetermined in their order and are fully controlled by the Trainer.
+
+Loop customization now enables a new level of control where also the two remaining `for` loops and more can be fully changed or replaced.
+
+Here is how the above training loop can be defined using the new Loop API:
+
+.. code-block:: python
+
+    class EpochLoop(Loop):
+        def advance(self):
+            i, batch = next(self.iterator)
+            loss = lightning_module.training_step(batch, i)
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+        def run(self, dataloader):
+            self.iterator = enumerate(dataloader)
+            while True:
+                try:
+                    self.advance()
+                except StopIteration:
+                    break
+
+Defining a loop with a class interface instead of hard-coding a raw Python for/while loop has several benefits:
+
+1. you can have full control over the data flow through loops
+2. you can add new loops and nest as many of them as they want
+3. if needed, the state of a loop can be saved and resumed
+4. new hooks can be injected at any point
+
+Which loops does Lightning have and how can they be changed?
+------------------------------------------------------------
+
+.. image:: ../_static/images/extensions/loops/replace-fit-loop.gif
+    :alt: Animation showing how to replace a loop on the Trainer
+
+
+The Trainer has four entry points for training, testing and inference, and each method corresponds to a main loop:
+
+.. list-table:: Trainer entry points and associated loops
+   :widths: 25 25 25
+   :header-rows: 1
+
+   * - Entry point
+     - Trainer attribute
+     - Loop class
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.fit`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.fit_loop`
+     - :class:`~pytorch_lightning.loops.fit_loop.FitLoop`
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.validate`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.validate_loop`
+     - :class:`~pytorch_lightning.loops.dataloader.evaluation_loop.EvaluationLoop`
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.test`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.test_loop`
+     - :class:`~pytorch_lightning.loops.dataloader.evaluation_loop.EvaluationLoop`
+   * - :meth:`~pytorch_lightning.trainer.trainer.Trainer.predict`
+     - :attr:`~pytorch_lightning.trainer.trainer.Trainer.predict_loop`
+     - :class:`~pytorch_lightning.loops.dataloader.prediction_loop.PredictionLoop`
```

Comment:
I really liked the FitLoop being decomposed as before into its part. I believe we might want to keep all this but at the bottom of the page.

More Technical Section.
...